packages/library/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt

```diff
@@ -99,7 +100,18 @@ object RealmObjectHelper {
         //  RealmObjectInterop and make cinterop operate on primitive values and native pointers
         //  only. This relates to the overall concern of having a generic path for getter/setter
         //  instead of generating a typed path for each type.
-        RealmInterop.realm_set_value(o, key, value, false)
+        try {
+            RealmInterop.realm_set_value(o, key, value, false)
+        }
+        // The catch block should catch specific Core exceptions and rethrow them as Kotlin exceptions.
+        // Core exceptions meaning might differ depending on the context, by rethrowing we can add some context related
+        // info that might help users to understand the exception.
+        catch (exception: RealmCoreNotInATransactionException) {
+            throw IllegalStateException(
+                "Cannot set `${obj.`$realm$TableName`}.$col` to `$value`: changing Realm data can only be done from inside a transaction.",
```

Comment:
```suggestion
                "Cannot set `${obj.`$realm$TableName`}.$col` to `$value`: changing Realm data can only be done on a live object from inside a write transaction. Frozen objects can be turned into live using the 'MutableRealm.findLatest(obj)' API.",
```

I think this is going to be a very common failure when first starting using the SDK, so we need to be very descriptive in the error IMO.